text/0007-performance-data-propagation.md

```diff
@@ -0,0 +1,179 @@
+
+External Tools and APIs will be able to then get this fully written `ProfilerTickData` at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a `Dictionary`, accessing its data will be as simple as accessing things by specified fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in `NetworkManager.LateUpdate()` for the MLAPI and `Transport.Send()` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data). Since MLAPI is responsible for collecting the data, and the data from transports will be generic and apply to all/most transports, MLAPI as  the owner of the collection of performance data will call `GetTransportGetData()` on the transports through an interface in order to get the data exactly when it needs it in the pipeline before propagating the data out externally. Whether or not the data is collected per transport can also be toggled by a runtime flag `profilerEnabled`.
+
+## Pseudo-code
+
+```cs
+namespace MLAPI
+{
+    class ProfilerConstants
+    {
+        public const string NumberOfServerRPCs = "numberOfServerRPCs";
+        public const string TickDuration = "tickDuration";
+        public const string NumberOfBuffMessagesOut = "numberOfBuffMessagesOut";
+        public const string NumberOfBuffMessagesIn = "numberOfBuffMessagesIn";
+        public const string NumberOfUnBuffMessagesOut = "numberOfUnBuffMessagesOut";
+        public const string NumberOfUnBuffMessagesIn = "numberOfUnBuffMessagesIn";
+        public const string NumberBytesSent = "numberBytesSent";
+        public const string NumberBytesReceived = "numberBytesReceived";
+    }
+
+    public class ProfilerTickData
+    {
+        public int tickID;
+
+        public readonly Dictionary<string, int> tickData = new Dictionary<string, int>();
+    }
+
+    public interface ITransportProfilerData
+    {
+        Dictionary<string, int> GetTransportGetData();
+    }
+}
+
+namespace MLAPI
+{
+    class NetworkManager
+    {
+        private static int tickID = 0;
+
+        public delegate void PerfomanceDataEventHandler(ProfilerTickData profilerData);
+
+        event PerfomanceDataEventHandler PerfomanceDataEvent;
+
+        void LateUpdate()
+        {
+            // ...
+            ProfilerTickData profilerData = new ProfilerTickData();
```

Comment:
Probably should use a pool, (impl I know)

Reply:
Will likely use a static pool where the receiver (third party) can't keep a reference or the reference is readonly